ballerina/introspection_validator_visitor.bal

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import graphql.parser;
+
+class IntrospectionValidatorVisitor {
+    *ValidatorVisitor;
+
+    private final boolean introspection;
+    private ErrorDetail[] errors;
+
+    isolated function init(boolean introspection) {
+        self.introspection = introspection;
+        self.errors = [];
+    }
+
+    public isolated function visitDocument(parser:DocumentNode documentNode, anydata data = ()) {
+        parser:OperationNode[] operations = documentNode.getOperations();
+        foreach parser:OperationNode operationNode in operations {
+            operationNode.accept(self);
+        }
+    }
+
+    public isolated function visitOperation(parser:OperationNode operationNode, anydata data = ()) {
+        foreach parser:SelectionNode selection in operationNode.getSelections() {
+            selection.accept(self, operationNode.getKind());
+        }
+    }
+
+    public isolated function visitField(parser:FieldNode fieldNode, anydata data = ()) {
+        if fieldNode.getName() == SCHEMA_FIELD || fieldNode.getName() == TYPE_FIELD {
+            if !self.introspection {
```

Comment:
We can just return from the `visitDocument` function if the introspection is disabled. No need to iterate through everything here.